### PR TITLE
V0

### PR DESCRIPTION
**Pull request by jean-michelet**
#31 #10 #28 #34

The aim of this PR is to produce enough features to release a v0 (backend only). 
I think it's simpler launch it this way, as the code changes a lot and it's easier to get feedback loop.

### CODE REVIEW

File: imports/ts-node.js
Comment (jean-michelet) [inline imports/ts-node.js:3]: @mcollina 

I have a warning from node when I try to use experimental loader.
The warning message advise to use this code instead, how would you use ts-node as part of a node command?
Example:
```
node --import ./imports/ts-node.js --env-file=.env ./scripts/seed-database.ts
```

File: src/routes/api/auth/index.ts
Comment (jean-michelet) [inline src/routes/api/auth/index.ts:41]: It does look more performant to put the roles in the jwt than doing an sql query at each request.
But it might have a drawback, e.g. I want a user not being an admin anymore, It could take some time to apply the changes (JWT expiration time).

**Review by mcollina — COMMENTED** [review]
good work! What's missing?

**Comment by jean-michelet** [comment]
> good work! What's missing?

### To do for a v0
- Use tsx
- I think we should add a pagination to the tasks
- I think we should be able to upload task thumbnails

People that might be give a feedback: @gurgunday @climba03003 @melroy89 @metcoder95 @melroy89

### Maybe for the future?
Initially, I wanted more features like:
- User API for Profile and Password Management
- Registration form with email confirmation

But, although useful from a pedagogical perspective, take the focus away from Fastify and its core plugins. Another problem is that more features leads to more work and maintenance, and I understand now that I will not have the help I expected to have in terms of contribution on this repo. So I think it's better to have a small, quality demo repo that I can commit to maintaining over the long term, than a more ambitious project that doesn't age well.

### Frontend
I also wanted to build a frontend with Fastify/Vite, there is [this PR](https://github.com/fastify/demo/pull/21) for it,  but I am not really confident with the TypeScript support of Fastify/Vite for now.

### Fastify typing
I would like to work on improving TypeScript on Fastify to improve this demo, to get rid of merging declaration, but I see a lot of different PRs and opinions on the subject. 

### CLI DX
There is also [this PR](https://github.com/fastify/fastify-cli/pull/743) from @climba03003 we should review to offer a better CLI experience.

**Comment by melroy89** [comment]
> People that might be give a feedback: @gurgunday @climba03003 @melroy89 @metcoder95 @melroy89

Twice me name 😆 . I'm just one person btw.

File: imports/ts-node.js
Comment (melroy89) [reply imports/ts-node.js:3]: Yea it used to be a big issue actually.. I needed to apply various workarounds in order to get loader working in Fastify:

```sh
FASTIFY_AUTOLOAD_TYPESCRIPT=1 VITEST=true node --no-warnings=ExperimentalWarning --loader ts-node/esm src/index.ts
```

I would like to see official support for `ts-node/esm`.

Comment (melroy89) [reply imports/ts-node.js:3]: Update: Ow it seems that `FASTIFY_AUTOLOAD_TYPESCRIPT=1 VITEST=true ` is no longer required anymore to get it working!? That is cool!

**Comment by jean-michelet** [comment]
Update: following a post on my LinkedIn, I have a lot of people reaching me to contribute on this repo. So there is a chance that we can add more features after this PR and getting more DX feedbacks from the community 🤞

File: imports/ts-node.js
Comment (climba03003) [reply imports/ts-node.js:3]: >  I'm now sponsoring you via GitHub!

I don't even notice it is available now. Applied long time ago and forgotten about it.
Thanks for your sponsor.

> Why wouldn't you recommend ts-node/esm?

I am also using `ts-node/esm` in my environment, but it seems like the loader do not get much attention from the `TypeScript` team.

File: src/plugins/external/session.ts
Comment (metcoder95) [inline src/plugins/external/session.ts:21]: I'd advice to also make this one configurable

File: src/routes/api/autohooks.ts
Comment (metcoder95) [inline src/routes/api/autohooks.ts:4]: I know this is from autoload experience, but it wouldn't be better to put this as a plugin that can be added on-demand; it seems quite counterintuitive the way is set.

Having the possibility of add this to the routes an implementer wants to protect seems to have better ergonomics than a soft check. wdyt?

**Review by metcoder95 — COMMENTED** [review]
Overall , LGTM, just left few comments

File: src/routes/api/autohooks.ts
Comment (jean-michelet) [reply src/routes/api/autohooks.ts:4]: Do you recommend to perform this check to every route that need authentication? A set of controllers?

If yes, everything under `/api` requires to be authenticated anyway, so I think it's much safer to have this check here and protect a large part of the application.

File: src/plugins/external/session.ts
Comment (jean-michelet) [reply src/plugins/external/session.ts:21]: Ok

File: src/routes/api/autohooks.ts
Comment (jean-michelet) [reply src/routes/api/autohooks.ts:4]: I can add a check here: [src/plugins/custom/authorization.ts](https://github.com/fastify/demo/pull/33/files#diff-00db5aed3aea31be2f7e173651cc17e6984158fc638ae2be571df33a65f56d3b)

Comment (metcoder95) [reply src/routes/api/autohooks.ts:4]: Don't have an opinion on that, I believe that it varies from project to project, org to org, developer to developer; nevertheless having an example of that that showcases how to protect on a per-route basis will be of help for the community as they might face this requirement and require for them to apply this logic.

Comment (jean-michelet) [reply src/routes/api/autohooks.ts:4]: I have already a per-route approach for authorization, but for authenticated users with roles, isn't it enough?

File: src/plugins/custom/authorization.ts
Comment (mcollina) [inline src/plugins/custom/authorization.ts:34]: why arent' we doing `request.isAdmin()` instead?

File: src/plugins/external/cookie.ts
Comment (mcollina) [inline src/plugins/external/cookie.ts:11]: Is this file needed at all? Can we do it together with `session`?

File: src/routes/api/auth/index.ts
Comment (mcollina) [inline src/routes/api/auth/index.ts:43]: I think this should use a transaction

File: src/routes/api/tasks/index.ts
Comment (mcollina) [inline src/routes/api/tasks/index.ts:42]: This is a bad practice and it's inefficient. Maybe use some custom SQL?

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:42]: ok

File: src/plugins/external/cookie.ts
Comment (jean-michelet) [reply src/plugins/external/cookie.ts:11]: We can

File: src/plugins/custom/authorization.ts
Comment (jean-michelet) [reply src/plugins/custom/authorization.ts:34]: Hum, now that I am thinking about it.
I am not sure.

File: src/routes/api/tasks/index.ts
Comment (climba03003) [reply src/routes/api/tasks/index.ts:42]: Maybe, 
```sql
SELECT *, COUNT(*) OVER() AS rowNum
FROM tasks
WHER ...
```
Not sure how it goes into knex.

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:42]: The problem is that it put the row (`total`) in every task:
```ts
 {
    id: 95,
    name: 'Task 3',
    author_id: 21,
    assigned_user_id: null,
    status: 'completed',
    created_at: 2024-10-18T06:02:07.000Z,
    updated_at: 2024-10-18T06:02:07.000Z,
    total: 89
  },
  {
    id: 94,
    name: 'Task 2',
    author_id: 20,
    assigned_user_id: 21,
    status: 'in-progress',
    created_at: 2024-10-18T06:02:07.000Z,
    updated_at: 2024-10-18T06:02:07.000Z,
    total: 89
  }
// ...
```

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts]: Can you take a look at [this commit](https://github.com/fastify/demo/pull/33/commits/da0fb0da7a710e079d0a86a2ae3eeeb5d90f919e) plz?

File: src/routes/api/auth/index.ts
Comment (jean-michelet) [reply src/routes/api/auth/index.ts:43]:
```ts
async function (request, reply) {
      const { username, password } = request.body

      return fastify.knex.transaction(async (trx) => {
        try {
          const user = await fastify
            .knex<Credentials>('users')
            .select('username', 'password')
            .where({ username })
            .first()

          if (user) {
            const isPasswordValid = await fastify.compare(
              password,
              user.password
            )
            if (isPasswordValid) {
              const roles = await fastify
                .knex<{ name: string }>('roles')
                .select('roles.name')
                .join('user_roles', 'roles.id', '=', 'user_roles.role_id')
                .join('users', 'user_roles.user_id', '=', 'users.id')
                .where('users.username', username)

              request.session.user = {
                username,
                roles: roles.map((role) => role.name)
              }

              await request.session.save()

              return { success: true }
            }
          }

          reply.status(401)
          return { message: 'Invalid username or password.' }
        } catch (error) {
          await trx.rollback()
          throw error
        }
      })
    }
```

How would you test that transaction rollback on error?
I don't know if there is a way to mock/override decorators.

@climba03003

Comment (climba03003) [reply src/routes/api/auth/index.ts:43]: I do see a way to monkey-patch `fastify.compare` and then throw at that point.

Comment (jean-michelet) [reply src/routes/api/auth/index.ts:43]: I can do this:
```

test('Transaction should rollback on error', async (t) => {
  mock.module(
    path.join(
      import.meta.dirname,
      '../../../../src/plugins/custom/scrypt.js'
    ),
    {
      defaultExport: fastifyPlugin(
        async (fastify: FastifyInstance) => {
          fastify.decorate('compare', (value: string, hash: string) => {
            throw new Error()
          })
        },
        {
          name: 'scrypt'
        }
      )
    }
  )

  const app = await build(t)

  const res = await app.inject({
    method: 'POST',
    url: '/api/auth/login',
    payload: {
      username: 'basic',
      password: 'password123$'
    }
  })

  assert.strictEqual(res.statusCode, 500)
})
```

But it means using experimental mock.module feature and it weirdly breaks coverage...

Plz, can you give me the solution if you know it?

Comment (jean-michelet) [reply src/routes/api/auth/index.ts:43]: Scrypt-coverage

Comment (jean-michelet) [reply src/routes/api/auth/index.ts:43]: Ignore my previous messages, I don't know why I tough decorators were immutable...

File: scripts/create-database.ts
Comment (Mathieuka) [inline scripts/create-database.ts:4]: Isn't the type Connection already inferred? If so, it's redundant

Comment (Mathieuka) [inline scripts/create-database.ts:4]: Isn't the type Connection already inferred? If so, it's redundant

File: scripts/migrate.ts
Comment (Mathieuka) [inline scripts/migrate.ts:9]:
```suggestion
interface PostgratorResult {
  rows: RowDataPacket[];
  fields: FieldPacket[];
}
```

File: scripts/migrate.ts
Comment (Mathieuka) [inline scripts/migrate.ts:1]:
```suggestion
import mysql, { Connection, RowDataPacket, FieldPacket } from 'mysql2/promise'
```

File: scripts/migrate.ts
Comment (Mathieuka) [inline scripts/migrate.ts:45]: It doesn't seem very useful to have 2 time `connection.end`, we can include it directly in the try-catch block in the finally block ? 

```
finally {
    try {
         await connection.end();
    } catch (error) {
         console.error(error);
    }
  }
```

File: scripts/seed-database.ts
Comment (Mathieuka) [inline scripts/seed-database.ts:27]:
```suggestion
if (tables.length) {
```

File: scripts/seed-database.ts
Comment (Mathieuka) [inline scripts/seed-database.ts:29]:
```suggestion
(row: Record<string, string>) => row[`Tables_in_${process.env.MYSQL_DATABASE}`]
```

File: scripts/seed-database.ts
Comment (Mathieuka) [inline scripts/seed-database.ts:54]: To improve type safety, instead of forcing the type via casting, can we create an use the `InsertResult` type ?

```
type InsertResult = {
  insertId: number;
}

const [userResult]: InsertResult[] = ...

const userId = (userResult)...
```

File: src/plugins/external/session.ts
Comment (Mathieuka) [inline src/plugins/external/session.ts:3]: Path aliasing in the future?

File: src/routes/api/tasks/index.ts
Comment (Mathieuka) [inline src/routes/api/tasks/index.ts:56]:
```suggestion
total: tasks.length ? Number(tasks[0].total) : 0
```

File: scripts/migrate.ts
Comment (jean-michelet) [reply scripts/migrate.ts:9]: Unfortunately this is enforced by the library I use...
```ts
  export interface BaseOptions {
    schemaTable?: string;
    validateChecksums?: boolean;
    migrationPattern?: string;
    newline?: string;
    execQuery?: (query: string) => Promise<{ rows: any[] }>;
    execSqlScript?: (sqlScript: string) => Promise<void>;
  }
```

Do you know how to extends that? Anyway, this not important for the scripts but if you know how to not use any tell me.

File: scripts/seed-database.ts
Comment (jean-michelet) [reply scripts/seed-database.ts:27]: You would perform a nullish check, although it should work here, this is semantically wrong imho.

File: scripts/seed-database.ts
Comment (jean-michelet) [reply scripts/seed-database.ts:54]: Here, `any` really is `[QueryResult, FieldPacket[]]` so it would not work, `mysql2` typing is pretty bad, they don't offer generics if I understand correctly. Maybe we could use `knex` here too. I will create an issue about it to investigate.

File: src/plugins/external/session.ts
Comment (jean-michelet) [reply src/plugins/external/session.ts:3]: Unlike frontend projects, I am not sure it would work well on a node.js server.

File: src/routes/api/tasks/index.ts
Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:61]: Same as my previous feedback.

**Comment by jean-michelet** [comment]
I should do a thorough review, especially to make sure the doc is up to date and that we are consistent in the plugin (code in general) organization.

But I think I can get most of your feedback already so I change the status to RFR!

File: src/routes/api/tasks/index.ts
Comment (jean-michelet) [inline src/routes/api/tasks/index.ts:248]: I realize mime type is not guessed from file content, if I send `one_line_csv.png`, it does work...

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:248]: Should I use `file-type`?
```ts
const fileBuffer = await file.toBuffer();
const fileType = await fileTypeFromBuffer(fileBuffer);
```

File: package.json
Comment (Fdawgs) [inline package.json:12]:
```suggestion
    "build": "rm -rf ./dist && tsc",
```

This won't work for Windows users using cmd or PowerShell.
You may need to either add a script to `scripts/` that cleans up dist and call that, or explicitly state in documentation for users to use Git Bash which comes with Git for Windows.

Comment (melroy89) [reply package.json:12]: Or just don't use windows 🤣🤭

File: src/plugins/external/session.ts
Comment (jean-michelet) [reply src/plugins/external/session.ts:3]: I have no opinion on the subject if others are ok, why not :+1:

**Review by gurgunday — APPROVED** [review]

**Review by climba03003 — APPROVED** [review]
It is a bit hard to review, as it contain nearly 50 files changes.
Overall, LGTM. Should merge it before adding more changes.

File: src/routes/api/tasks/index.ts
Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:247]: No opinion about this @climba03003 @gurgunday?

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:247]: After this feedback, I will merge.

**Comment by jean-michelet** [comment]
> It is a bit hard to review, as it contain nearly 50 files changes.

Sorry about that, I really wanted to achieve the main backend features.

File: src/routes/api/tasks/index.ts
Comment (Fdawgs) [reply src/routes/api/tasks/index.ts:247]: I'd say yes, as it's a security risk. It's something I checked in one of my APIs: https://github.com/Fdawgs/docsmith/blob/main/src/routes/pdf/html/index.js#L32-L51

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:247]: Great the link, this is a good reason for educating about `addContentTypeParser` utility.
But don't you think we should mention this security risk in the plugin documentation?

Comment (gurgunday) [reply src/routes/api/tasks/index.ts:247]: There could be a mention, yeah

But it's ultimately the same idea of never trusting a client without input validation

Comment (jean-michelet) [reply src/routes/api/tasks/index.ts:247]: But do we offer this kind of validation, or does a package we use offer this kind of validation?

Doesn't look like we recommend real validation for content file by reading the doc.

Comment (climba03003) [reply src/routes/api/tasks/index.ts:247]: I am not very strong opinion on inspecting the file content for the content-type, most of the application only use `.ext` to `content-type` (e.g. Reverse Proxy, Browser, OS).
If it is an input file (batch processing), you should validate the content inside instead of the guessed `content-type`.
If it is an image file or the others, using wrong content-type does make changes (malicious code can still be hide inside the correct content-type and execute when parsing).

**Review by metcoder95 — APPROVED** [review]

**Comment by jean-michelet** [comment]
OK, I think I am gonna merge and will create issues to fix the existing or add new features.

It could be great to have more reviews, especially on the security aspect.

**Pull request merged**
